pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,527 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memberlist
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/golang/groupcache/consistenthash"
+	"github.com/hashicorp/memberlist"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha2"
+	crdlister "antrea.io/antrea/pkg/client/listers/crd/v1alpha2"
+	"antrea.io/antrea/pkg/util/k8s"
+)
+
+const (
+	controllerName = "MemberListCluster"
+	// Set resyncPeriod to 0 to disable resyncing.
+	resyncPeriod time.Duration = 0
+	// set default virtual node replicas num of consistent hash
+	defaultVirtualNodeReplicas = 3
```

Comment:
In your test, is 3 enough to distribute the egresses averagely? We need to consider the number seriously as changing it would cause upgrade issue in the future.

Reply:
I have test 3 nodes with 3 replicas which lead to an average distribute of a different scale of egresses, 
maybe set replicas to 50 is a recommended practice, reference  to https://github.com/golang/groupcache/issues/29

replicas = 1
```
//=== RUN   TestCluster_ShouldSelectByConstentHash/select_node_from_alive_nodes-3-egresses
//=== CONT  TestCluster_ShouldSelectByConstentHash
//cluster_test.go:612: Node: node1, egressNum: 1
//cluster_test.go:612: Node: node2, egressNum: 1
//cluster_test.go:612: Node: node3, egressNum: 1
//=== RUN   TestCluster_ShouldSelectByConstentHash/select_node_from_alive_nodes-10-egresses
//=== CONT  TestCluster_ShouldSelectByConstentHash
//cluster_test.go:612: Node: node1, egressNum: 4
//cluster_test.go:612: Node: node2, egressNum: 2
//cluster_test.go:612: Node: node3, egressNum: 4
//=== RUN   TestCluster_ShouldSelectByConstentHash/select_node_from_alive_nodes-100-egresses
//=== CONT  TestCluster_ShouldSelectByConstentHash
//cluster_test.go:612: Node: node1, egressNum: 44
//cluster_test.go:612: Node: node2, egressNum: 43
//cluster_test.go:612: Node: node3, egressNum: 13
//=== RUN   TestCluster_ShouldSelectByConstentHash/select_node_from_alive_nodes-1000-egresses
//=== CONT  TestCluster_ShouldSelectByConstentHash
//cluster_test.go:612: Node: node1, egressNum: 474
//cluster_test.go:612: Node: node2, egressNum: 408
//cluster_test.go:612: Node: node3, egressNum: 118
//=== RUN   TestCluster_ShouldSelectByConstentHash/select_node_from_alive_nodes-10000-egresses
//=== CONT  TestCluster_ShouldSelectByConstentHash
//cluster_test.go:612: Node: node1, egressNum: 5284
//cluster_test.go:612: Node: node2, egressNum: 3390
//cluster_test.go:612: Node: node3, egressNum: 1326
//--- PASS: TestCluster_ShouldSelectByConstentHash (0.09s)
```

replicas = 3

```
//=== RUN   TestCluster_ShouldSelectByConstentHash/select_node_from_alive_nodes-3-egresses
//=== CONT  TestCluster_ShouldSelectByConstentHash
//cluster_test.go:612: Node: node1, egressNum: 0
//cluster_test.go:612: Node: node2, egressNum: 2
//cluster_test.go:612: Node: node3, egressNum: 1
//=== RUN   TestCluster_ShouldSelectByConstentHash/select_node_from_alive_nodes-10-egresses
//=== CONT  TestCluster_ShouldSelectByConstentHash
//cluster_test.go:612: Node: node1, egressNum: 1
//cluster_test.go:612: Node: node2, egressNum: 5
//cluster_test.go:612: Node: node3, egressNum: 4
//=== RUN   TestCluster_ShouldSelectByConstentHash/select_node_from_alive_nodes-100-egresses
//=== CONT  TestCluster_ShouldSelectByConstentHash
//cluster_test.go:612: Node: node1, egressNum: 30
//cluster_test.go:612: Node: node2, egressNum: 33
//cluster_test.go:612: Node: node3, egressNum: 37
//=== RUN   TestCluster_ShouldSelectByConstentHash/select_node_from_alive_nodes-1000-egresses
//=== CONT  TestCluster_ShouldSelectByConstentHash
//cluster_test.go:612: Node: node1, egressNum: 348
//cluster_test.go:612: Node: node2, egressNum: 317
//cluster_test.go:612: Node: node3, egressNum: 335
//=== RUN   TestCluster_ShouldSelectByConstentHash/select_node_from_alive_nodes-10000-egresses
//=== CONT  TestCluster_ShouldSelectByConstentHash
//cluster_test.go:612: Node: node1, egressNum: 3318
//cluster_test.go:612: Node: node2, egressNum: 3469
//cluster_test.go:612: Node: node3, egressNum: 3213
//--- PASS: TestCluster_ShouldSelectByConstentHash (0.10s)
```

replicas = 50

```
//=== RUN   TestCluster_ShouldSelectByConstentHash/select_node_from_alive_nodes-3-egresses
//=== CONT  TestCluster_ShouldSelectByConstentHash
//cluster_test.go:612: Node: node1, egressNum: 1
//cluster_test.go:612: Node: node2, egressNum: 2
//cluster_test.go:612: Node: node3, egressNum: 0
//=== RUN   TestCluster_ShouldSelectByConstentHash/select_node_from_alive_nodes-10-egresses
//=== CONT  TestCluster_ShouldSelectByConstentHash
//cluster_test.go:612: Node: node1, egressNum: 1
//cluster_test.go:612: Node: node2, egressNum: 4
//cluster_test.go:612: Node: node3, egressNum: 5
//=== RUN   TestCluster_ShouldSelectByConstentHash/select_node_from_alive_nodes-100-egresses
//=== CONT  TestCluster_ShouldSelectByConstentHash
//cluster_test.go:612: Node: node1, egressNum: 30
//cluster_test.go:612: Node: node2, egressNum: 32
//cluster_test.go:612: Node: node3, egressNum: 38
//=== RUN   TestCluster_ShouldSelectByConstentHash/select_node_from_alive_nodes-1000-egresses
//=== CONT  TestCluster_ShouldSelectByConstentHash
//cluster_test.go:612: Node: node1, egressNum: 345
//cluster_test.go:612: Node: node2, egressNum: 297
//cluster_test.go:612: Node: node3, egressNum: 358
//=== RUN   TestCluster_ShouldSelectByConstentHash/select_node_from_alive_nodes-10000-egresses
//=== CONT  TestCluster_ShouldSelectByConstentHash
//cluster_test.go:612: Node: node1, egressNum: 3437
//cluster_test.go:612: Node: node2, egressNum: 2930
//cluster_test.go:612: Node: node3, egressNum: 3633
//--- PASS: TestCluster_ShouldSelectByConstentHash (0.10s)
```

Reply:
done